lookout/style/format/analyzer.py

```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
     @with_uasts_and_contents
     def train(cls, ptr: ReferencePointer, config: dict, data_request_stub: DataStub,
               **data) -> AnalyzerModel:
+        """
+        Train a model given the files available.
+
+        :param url: url of the repository
+        :param commit: hash of the commit that we analyze
+        :param config: configuration dict
+        :param data: data sent by the lookout server
+        :return: a modelforge.Model containing the learned rules, per language.
+        """
         cls.log.info("train %s %s %s", ptr.url, ptr.commit, data)
         files = data["files"]
+        language_to_files = defaultdict(list)
         for file in files:
+            if not len(file.uast.children):
+                continue
+            # TODO: replace this temporary hack when langauge fields in File objects are fixed
+            extension = file.path.rsplit(".", 1)[1]
+            if extension != "js":
+                continue
+            language_to_files["javascript"].append(file)
             cls.log.info("%s %s %d", file.path, file.language, len(file.uast.children))
 
-        """
-        Plan:
-
-        1. Load all the File-s - they are streamed and we need to read them all into a list
-        2. Sort by language
-        3. Loop per language group:
-        4. extract_features()
-        5. train the base model - tree or forest
-        6. train the rules
-        7. generate the Modelforge model
-        """
+        model = FormatModel().construct(cls, ptr)
 
-        return FormatModel().construct(cls, ptr)
+        for language, files in language_to_files.items():
+            cls.log.info("processing %s files", language)
+            cls.log.info("extracting features from %d files", len(files))
+            fe = FeatureExtractor(language=language, siblings_window=5, parents_depth=2)
+            X, y = fe.extract_features(files)
+            if X.shape[0] < cls.LOWER_BOUND_INSTANCES:
+                cls.log.warning("did not train a model for language %s: not sufficiently many "
+                                "instances (%d)" % (language, X.shape[0]))
```

Comment:
We seem to miss `continue` here

Reply:
`"skipped %s: too few samples (%d < %d)" % (file.path, X.shape[0], min_samples)`